File: codesignal/interview_practice/1.data structures/1.arrays/1.firstDuplicate.js
// Given an array a that contains only numbers in the range from 1 to a.length,
// find the first duplicate number for which the second occurrence has the minimal index.
// In other words, if there are more than 1 duplicated numbers, return the number for which
// the second occurrence has a smaller index than the second occurrence of the other number does.
// If there are no such elements, return -1.

class Solution {
  firstDuplicate(a) {
    let minIndex = a.length;
    let minNumber = -1;
    for (let i = 0; i < a.length; i++) {
      for (let j = i + 1; j < a.length; j++) {
        if (a[i] === a[j]) {
          if (j < minIndex) {
            minIndex = j;
            minNumber = a[i];
          }
        }
      }
    }
    return minNumber;
  }
}

class Solution {
  firstDuplicate(a) {
    r = new Set();
    for (let i = 0; i < a.length; i++) {
      if (r.has(a[i])) return a[i];
      r.add(a[i]);
    }
    return -1;
  }
}

class Solution {
  firstDuplicate(a) {
    duplicates = {};
    for (let i = 0; i < a.length; i++) {
      if (duplicates[a[i]]) return a[i];
      else duplicates[a[i]] = 1;
    }
    return -1;
  }
}

File: codesignal/interview_practice/1.data structures/1.arrays/1.firstDuplicate.cpp
// # Given an array a that contains only numbers in the range from 1 to a.length, 
// find the first duplicate number for which the second occurrence has the minimal index. 
// In other words, if there are more than 1 duplicated numbers, return the number for which 
// the second occurrence has a smaller index than the second occurrence of the other number does. 
// If there are no such elements, return -1.

#include <unordered_set>

class Solution {
public:
  int firstDuplicate(std::vector<int> a) {
    std::unordered_set<int> s;
    for (int i = 0; i < a.size(); i++) {
      if (s.find(a[i]) != s.end()) {
        return a[i];
      }
      s.insert(a[i]);
    }
    return -1;
  }
};

class Solution {
public:
  int firstDuplicate(std::vector<int> a) {
    for(int i=0; i < a.size(); i++){
      int t = abs(a[i]);
      if(a[t - 1] < 0)
        return t;
      a[t - 1] = - a[t - 1];
    }
    return -1;
  }
};